client/shipping-label/state/reducer.js

```diff
@@ -328,25 +329,39 @@ reducers[ OPEN_ADD_ITEM ] = ( state ) => {
 	return {
 		...state,
 		showAddItemDialog: true,
+		addedItems: {},
 	};
 };
 
 reducers[ CLOSE_ADD_ITEM ] = ( state ) => {
 	return {
 		...state,
-		movedItemIndex: -1,
 		showAddItemDialog: false,
 	};
 };
 
-reducers[ SET_ADDED_ITEM ] = ( state, { sourcePackageId, movedItemIndex } ) => {
+reducers[ SET_ADDED_ITEM ] = ( state, { sourcePackageId, movedItemIndex, added } ) => {
 	return {
 		...state,
-		sourcePackageId,
-		movedItemIndex,
+		addedItems: { ...state.addedItems,
+			[ sourcePackageId ]: added
+				? { ...state.addedItems[ sourcePackageId ], [ movedItemIndex ]: true }
+				: _.omit( state.addedItems[ sourcePackageId ], movedItemIndex )
+		},
 	};
 };
 
+reducers[ ADD_ITEMS ] = ( state, { targetPackageId } ) => {
+	// For each origin package
+	_.each( state.addedItems, ( inPackage, originPackageId ) => {
+		// Move items in reverse order of index, to maintain validity as items are removed
```

Comment:
This was the bug with the original implementation in this PR. Adding example in expanded comment, but let me know if you find it to be a bad practice / code smell.